From d697cdaf48b347d5da39ba015a3c87ece52b9fa1 Mon Sep 17 00:00:00 2001
From: Pino Toscano <ptoscano@redhat.com>
Date: Thu, 13 Nov 2014 05:01:54 -0500
Subject: [PATCH 070/118] customize: firstboot: make sure to run Linux scripts
 only once

If a script does not finish, hangs, etc, it would be executed again at
the next boot, since the injected firstboot.sh helper removes it only
after it finished.

Before executing a script, move it to another internal subdirectory
(scripts-done) and execute it from there, so it is not run again by
firstboot.sh.  The downside is that now scripts are executed only once
at all, so if a script fails it will not be attempted at the next boot.

Also, remove all the files found in scripts-done, as they have been run
(or at least attempted) in a previous boot.

This fixes RHBZ#1159651.
(cherry picked from commit f8ed15462fbb03c5b19972361f2a2e6fed4c5f02)
---
 mllib/firstboot.ml | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/mllib/firstboot.ml b/mllib/firstboot.ml
index 9e4c7b6..ea79cb8 100644
--- a/mllib/firstboot.ml
+++ b/mllib/firstboot.ml
@@ -43,6 +43,7 @@ let firstboot_sh = sprintf "\
 ### END INIT INFO
 
 d=%s/scripts
+d_done=%s/scripts-done
 logfile=~root/virt-sysprep-firstboot.log
 
 echo \"$0\" \"$@\" 2>&1 | tee $logfile
@@ -50,16 +51,20 @@ echo \"Scripts dir: $d\" 2>&1 | tee $logfile
 
 if test \"$1\" = \"start\"
 then
+  mkdir -p $d_done
   for f in $d/* ; do
     if test -x \"$f\"
     then
+      # move the script to the 'scripts-done' directory, so it is not
+      # executed again at the next boot
+      mv $f $d_done
       echo '=== Running' $f '===' 2>&1 | tee $logfile
-      $f 2>&1 | tee $logfile
-      rm -f $f
+      $d_done/$(basename $f) 2>&1 | tee $logfile
     fi
   done
+  rm -f $d_done/*
 fi
-" firstboot_dir
+" firstboot_dir firstboot_dir
 
 let firstboot_service = sprintf "\
 [Unit]
-- 
2.7.4

